src/main/scala/is/hail/keytable/KeyTable.scala

```diff
@@ -272,72 +223,63 @@ case class KeyTable(@transient hc: HailContext, @transient rdd: RDD[(Annotation,
-    if (keySignature != other.keySignature)
+    if (keyNames.length != other.keyNames.length || !(keyFields.map(_.typ) sameElements other.keyFields.map(_.typ)))
       fatal(
-        s"""Key signatures must be identical.
-           |Left signature: ${ keySignature.toPrettyString(compact = true) }
-           |Right signature: ${ other.keySignature.toPrettyString(compact = true) }""".stripMargin)
+        s"""Both key tables must have the same number of keys and the types of keys must be identical. Order matters.
+           |Left signature: ${ TStruct(keyFields).toPrettyString(compact = true) }
+           |Right signature: ${ TStruct(other.keyFields).toPrettyString(compact = true) }""".stripMargin)
 
-    val overlappingFields = valueNames.toSet.intersect(other.valueNames.toSet)
+    val overlappingFields = fieldNames.toSet.intersect(other.fieldNames.toSet) -- keyNames -- other.keyNames
     if (overlappingFields.nonEmpty)
       fatal(
-        s"""Fields that are not keys cannot be present in both key-tables.
+        s"""Columns that are not keys cannot be present in both key tables.
            |Overlapping fields: ${ overlappingFields.mkString(", ") }""".stripMargin)
 
-    joinType match {
-      case "left" => leftJoin(other)
-      case "right" => rightJoin(other)
-      case "inner" => innerJoin(other)
-      case "outer" => outerJoin(other)
-      case _ => fatal("Invalid join type specified. Choose one of `left', `right', `inner', `outer'")
-    }
-  }
-
-  def leftJoin(other: KeyTable): KeyTable = {
-    require(keySignature == other.keySignature)
+    val mergeFields = other.fields.filterNot(fd => other.keyNames.contains(fd.name))
+    val mergeIndices = mergeFields.map(_.index)
 
-    val (newValueSignature, merger) = valueSignature.merge(other.valueSignature)
-    val newRDD = rdd.leftOuterJoin(other.rdd).map { case (k, (vl, vr)) => (k, merger(vl, vr.orNull)) }
+    val newSignature = TStruct((fields ++ mergeFields).map(fd => (fd.name, fd.typ)): _*)
 
-    KeyTable(hc, newRDD, keySignature, newValueSignature)
-  }
-
-  def rightJoin(other: KeyTable): KeyTable = {
-    require(keySignature == other.keySignature)
-
-    val (newValueSignature, merger) = valueSignature.merge(other.valueSignature)
-    val newRDD = rdd.rightOuterJoin(other.rdd).map { case (k, (vl, vr)) => (k, merger(vl.orNull, vr)) }
-
-    KeyTable(hc, newRDD, keySignature, newValueSignature)
-  }
+    val size1 = nFields
+    val targetSize = newSignature.size
 
-  def outerJoin(other: KeyTable): KeyTable = {
-    require(keySignature == other.keySignature)
-
-    val (newValueSignature, merger) = valueSignature.merge(other.valueSignature)
-    val newRDD = rdd.fullOuterJoin(other.rdd).map { case (k, (vl, vr)) => (k, merger(vl.orNull, vr.orNull)) }
-
-    KeyTable(hc, newRDD, keySignature, newValueSignature)
-  }
+    val merger = (a1: Annotation, a2: Annotation) => {
+      if (a1 == null && a2 == null)
+        Annotation.empty
```

Comment:
You should return an array of nulls in this case.  Just pull up the construction of `result`.